src/common/resolve.rs

```diff
@@ -52,25 +53,33 @@ pub(super) fn resolve_launch_and_shell(
 }
 
 pub(crate) fn resolve_target_user_and_group(
-    target_user_name_or_id: &Option<String>,
-    target_group_name_or_id: &Option<String>,
+    target_user_name_or_id: &Option<SudoString>,
+    target_group_name_or_id: &Option<SudoString>,
     current_user: &User,
 ) -> Result<(User, Group), Error> {
     // resolve user name or #<id> to a user
-    let mut target_user =
-        match NameOrId::parse(target_user_name_or_id.as_deref().unwrap_or_default()) {
-            Some(NameOrId::Name(name)) => User::from_name(name)?,
-            Some(NameOrId::Id(uid)) => User::from_uid(uid)?,
-            _ => None,
-        };
+    let mut target_user = target_user_name_or_id
+        .as_ref()
+        .and_then(|input| {
+            match NameOrId::parse(input)? {
+                NameOrId::Name(name) => User::from_name(name.as_cstr()),
+                NameOrId::Id(uid) => User::from_uid(uid),
+            }
+            .transpose()
+        })
+        .transpose()?;
```

Comment:
this double `transpose` feels sub-optimal. if someone has a suggestion, I'm all ears

Reply:
I played for a while with this and I think it would be possible to factor this whole transpose magic into a new function like this:
```rust
fn resolve_name_or_id<T, I: FromStr, E>(
    input: &Option<SudoString>,
    from_name: impl Fn(&std::ffi::CStr) -> Result<Option<T>, E>,
    from_id: impl Fn(I) -> Result<Option<T>, E>,
) -> Result<Option<T>, E> {
    match input.as_ref().and_then(NameOrId::parse) {
        Some(name_or_id) => match name_or_id {
            NameOrId::Name(name) => from_name(name.as_cstr()),
            NameOrId::Id(id) => from_id(id),
        },
        None => Ok(None),
    }
}
```
So you can do
```rust
// resolve user name or #<id> to a user
let mut target_user =
    resolve_name_or_id(target_user_name_or_id, User::from_name, User::from_uid)?;

// resolve group name or #<id> to a group
let mut target_group =
    resolve_name_or_id(target_group_name_or_id, Group::from_name, Group::from_gid)?;
```

Not sure if its more-optimal or not but it encapsulates the resolution logic in its own function.

PD: I'm also not sure if it is right or not :rofl:

Reply:
applied your refactor suggestion